File: SampleAppObjC/sources/viewControllers/CustomSDKViewController.m
//
//  CustomSDKViewController.m
//  SampleAppObjC
//
//  Copyright © 2017-2018 BEAR SAS. All rights reserved.
//

#import "CustomSDKViewController.h"
#import <UserNotifications/UserNotifications.h>
#import "Cache.h"
#import "HistoryViewController.h"

#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@interface CustomSDKViewController ()

@property (weak, nonatomic) IBOutlet UIButton *startStopBtn;
@property (weak, nonatomic) IBOutlet UIButton *flashBtn;
@property (weak, nonatomic) IBOutlet UIButton *screenshotBtn;
@property (weak, nonatomic) IBOutlet UIView *splashScreen;
@property (weak, nonatomic) IBOutlet UIButton *historyBtn;

@end

@implementation CustomSDKViewController

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self initDefault];
    }
    return self;
}

-(void)initDefault {
    _enableHistory = NO;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.delegate = self;
    [_historyBtn setHidden:!_enableHistory];
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (self.navigationController != nil) {
        [self.navigationController setNavigationBarHidden:YES animated:animated];
    }
}

-(void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [_splashScreen setHidden:YES];
}

- (IBAction)dismiss {
    if (self.navigationController != nil) {
        [self.navigationController popViewControllerAnimated:YES];
    } else {
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}

- (IBAction)startStopTapped {
    if (self.handler.scannerState == BearScannerStateIdle) {
        [self.handler startScanning];
    } else if (self.handler.scannerState == BearScannerStateScanning) {
        [self.handler stopScanning];
    }
}

- (IBAction)enableDisableFlash {
    if (!self.handler.isFlashEnabled) {
        [self.flashBtn setSelected: [self.handler enableFlash]];
    } else {
        [self.handler disableFlash];
        [_flashBtn setSelected:NO];
    }
}

- (IBAction)showARSceneWithoutTracking {
    [self.handler showARSceneWithoutTrackingWithMarkerId:226620];
}

- (IBAction)takeScreenshot {
    UIImage* image = [self.handler takeScreenshot];
    if (image != nil) {
        [self presentShareWithScreenshot:image];
    }
}

- (IBAction)cleanView {
    [self.handler cleanView];
}

- (IBAction)registerNotifications {
    if (SYSTEM_VERSION_LESS_THAN(@"10.0")) {
        UIUserNotificationSettings* settings = [UIUserNotificationSettings
                                                settingsForTypes:UIUserNotificationTypeAlert | UIUserNotificationTypeBadge | UIUserNotificationTypeSound
                                                categories:nil];
        [[UIApplication sharedApplication] registerUserNotificationSettings:settings];
        [[UIApplication sharedApplication] registerForRemoteNotifications];
    } else {
        [[UNUserNotificationCenter currentNotificationCenter]
         requestAuthorizationWithOptions:UNAuthorizationOptionAlert | UNAuthorizationOptionBadge | UNAuthorizationOptionSound
         completionHandler:^(BOOL granted, NSError * _Nullable error) {
             if (!granted) return;
             dispatch_async(dispatch_get_main_queue(), ^{
                 [[UIApplication sharedApplication] registerForRemoteNotifications];
             });
         }];
    }
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if (![[segue destinationViewController] isKindOfClass:[HistoryViewController class]]) return;
    HistoryViewController* vc = (HistoryViewController*)[segue destinationViewController];
    vc.hidePreloadBtn = YES;
}

-(void)presentShareWithScreenshot:(UIImage*)screenshot {
    UIActivityViewController* activityVC = [[UIActivityViewController alloc] initWithActivityItems:@[screenshot] applicationActivities:nil];
    
    activityVC.modalPresentationStyle = UIModalPresentationPopover;
    if (activityVC.popoverPresentationController != nil) {
        activityVC.popoverPresentationController.sourceRect = _screenshotBtn.frame;
        activityVC.popoverPresentationController.sourceView = self.view;
    }
    [self presentViewController:activityVC animated:true completion:nil];
}

// MARK: BearDelegate
- (void)assetClickedWith:(NSInteger)assetId {
    NSLog(@"assetClickedWith = %li", (long)assetId);
}

- (void)markerNotRecognized {
    NSLog(@"markerNotRecognized");
}

- (void)recognizedMarkerWithId:(NSInteger)markerId assetIds:(NSArray<NSNumber *> * _Nonnull)assetIds {
    NSLog(@"recognizedMarkerWithId = %ld, assetIds = %@", (long)markerId, assetIds);
    [Cache cacheRecognizedMarkerId:markerId];
}

- (void)scannerStateChanged:(enum BearScannerState)state {
    switch (state) {
        case BearScannerStateIdle:
            NSLog(@"BearScannerStateIdle");
            [_flashBtn setSelected:NO];
            [_startStopBtn setSelected:NO];
            break;
        case BearScannerStateScanning:
            NSLog(@"BearScanStateScanning");
            [_startStopBtn setSelected:YES];
            break;
        case BearScannerStateProcessing:
            NSLog(@"BearScanStateProcessing");
            [_flashBtn setSelected:NO];
            break;
        case BearScannerStateRendering:
            NSLog(@"BearScannerStateRendering");
            [_startStopBtn setSelected:NO];
            break;
    }
}

- (void)didFailWithError:(BearError* _Nonnull)error {
    NSLog(@"didFailWithError %@", error);
    UIAlertController* alertVC = [UIAlertController alertControllerWithTitle:error.shortDescription
                                                                     message:error.reason preferredStyle:UIAlertControllerStyleAlert];
    [alertVC addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alertVC animated:YES completion:nil];
}

@end

File: SampleApp/sources/viewControllers/CustomSDKViewController.swift
//
//  CustomSDKViewController.swift
//  SampleApp
//
//  Copyright © 2017 BEAR SAS. All rights reserved.
//

import Foundation
import BearSDK
import UserNotifications

final class CustomSDKViewController: BearViewController {
    
    @IBOutlet weak var startStopBtn: UIButton!
    @IBOutlet weak var flashBtn: UIButton!
    @IBOutlet weak var screenshotBtn: UIButton!
    @IBOutlet weak var splashScreen: UIView!
    @IBOutlet weak var historyBtn: UIButton!
    var enableHistory: Bool = false
    var hideNavBar: Bool = true
    
    override func viewDidLoad() {
        super.viewDidLoad()
        delegate = self
        historyBtn.isHidden = !enableHistory
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if hideNavBar {
            navigationController?.setNavigationBarHidden(true, animated: true)
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        splashScreen.isHidden = true
    }
    
    @IBAction func dismiss() {
        if let nav = navigationController {
            nav.popViewController(animated: true)
        } else {
            dismiss(animated: true, completion: nil)
        }
    }
    
    @IBAction func startStopTapped() {
        if handler.scannerState == .idle {
            handler.startScanning()
        } else if handler.scannerState == .scanning {
            handler.stopScanning()
        }
    }
    
    @IBAction func enableDisableFlash() {
        if !handler.isFlashEnabled {
            flashBtn.isSelected = handler.enableFlash()
        } else {
            handler.disableFlash()
            flashBtn.isSelected = false
        }
    }
    
    @IBAction func showARSceneWithoutTracking() {
        handler.showARSceneWithoutTracking(withMarkerId: 226620)
    }
    
    @IBAction func takeScreenshot() {
        if let screenshot = handler.takeScreenshot {
            presentShare(withScreenshot: screenshot)
        }
    }
    
    @IBAction func cleanView() {
        handler.cleanView()
    }
    
    @IBAction func registerNotifications() {
        if #available(iOS 10.0, *) {
            UNUserNotificationCenter.current().requestAuthorization(options: [.badge, .alert , .sound]) { (granted, error) in
                guard granted else { return }
                DispatchQueue.main.async {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            }
        } else {
            let settings = UIUserNotificationSettings(types: [.alert, .sound, .badge], categories: nil)
            UIApplication.shared.registerUserNotificationSettings(settings);
            UIApplication.shared.registerForRemoteNotifications();
        }
    }
}

extension CustomSDKViewController: BearDelegate {
    func recognizedMarker(withId markerId: Int, assetIds: [Int]) {
        debugPrint("called \(#function), markerId = \(markerId), assetIds = \(assetIds)")
        cacheRecognized(markerId: markerId)
    }
    
    func scannerStateChanged(_ state: BearScannerState) {
        switch state {
        case .idle:
            debugPrint("idle")
            flashBtn.isSelected = false
            startStopBtn.isSelected = false
            return
        case .scanning:
            debugPrint("scanning")
            startStopBtn.isSelected = true
            return
        case .processing:
            debugPrint("processing")
            flashBtn.isSelected = false
            return
        case .rendering:
            debugPrint("rendering")
            startStopBtn.isSelected = false
            return
        }
    }
    
    func markerNotRecognized() {
        debugPrint("called \(#function)")
    }
    
    func assetClicked(with assetId: Int) {
        debugPrint("called \(#function), assetId = \(assetId)")
    }
    
    func reachabilityStatusChanged(_ reachable: Bool) {
        debugPrint("called \(#function), reachable = \(reachable)")
    }
    
    func didFail(withError error: BearError) {
        debugPrint("called \(#function)")
        showAlert(withTitle: error.shortDescription, message: error.reason)
    }
    
    private func showAlert(withTitle title: String, message: String?) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        present(alert, animated: true, completion: nil)
    }
    
    private func presentShare(withScreenshot screenshot: UIImage) {
        
        let activityVC = UIActivityViewController(activityItems: [screenshot], applicationActivities: nil)
        
        activityVC.modalPresentationStyle = .popover
        activityVC.popoverPresentationController?.sourceRect = screenshotBtn.frame
        activityVC.popoverPresentationController?.sourceView = view
        
        present(activityVC, animated: true, completion: nil)
    }
}

